docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+
+### Authorization ###
+
+For authorization, a system tailored to Heapster directly would be more useful:
+
+```go
+// This is basically just core.HistoricalKey from the Oldtimer PR,
+// renamed to be more generic
+type MetricKey struct {
+    // One of the core.MetricSetType* labels
+    ObjectType string
+
+    NodeName string
+    NamespaceName string
+    PodName string
+    ContainerName string
+    PodId string
+}
+
+type HeapsterQuerySource string
+var (
+    QuerySourceModel HeapsterQuerySource = "model"
+    QuerySourceHistorical HeapsterQuerySource = "historical"
+)
+
+type Authorizer interface {
+    // AuthorizeCreation authorizes creation of metrics with the given prefix
+    // (if any) and the given names at the given keys.
+    AuthorizeCreation(user kauthuser.Info, keys []MetricKey, prefix string, metricNames []string) error 
+
+    // AuthorizeQuery authorizes querying of metrics at the given keys with the
+    // given names from the given source.
+    AuthorizeQuery(user kauthuser.Info, source HeapsterQuerySource, keys []MetricKey, metricNames []string) error
+}
+```
+
+The `Authorizer` interface allows authorizing both queries against Heapster
+and requests to create new metrics entries.
+
+For queries, `AuthorizeQuery` takes user information, a source from which the
+information will be queried (the model or the historical source), and a set of
+keys and metric names.
+
+The authorizer may not assume a relationship between keys and metric names --
+it must assume all metric names given may be pushed for all metric keys (see
+below for the rationale for passing metric names).
+
+The authorizer will then take the given information, and return an error if
+the given user is not authorized to perform the given query.
+
+For creation of metrics, `AuthorizeCreation` takes much the same parameters
+as `AuthorizeQuery` (except for the `source` parameter), with the addition
+of the `prefix` parameter.  If the prefix parameter is given, when checking
+metric names, the authorizer should first prepend the prefix to the metric
+names.
+
+Passing keys to the authorizer allows for fine-grained scope decisions to be
+made -- some users may be only able to query metrics in their own namespace,
+while others may be able to query any metrics.  Some users may be able to push
+metrics only for their own namespace, or even for their own pod, while others
+may be able to push metrics for any key in Heapster.
+
+Passing in metric names allows the authorizer to make complex decisions about
+which users are allowed to push all which metrics.  For instance, an authorizer
+could restrict certain users to pushing metrics with a certain prefix, while
+allowing another user to push metrics without a prefix, as long as they were
+not one of the built-in metrics.  On the query side, this could allow
+metric-level restrictions on queries, such as only allowing certain users to
+read a certain set of custom metrics, but not have access to built-in metrics.
```

Comment:
The decision is mostly on the "prefix" vs "no prefix" -- it could allow for a extension to push metrics where certain applications are trusted to push metrics without the "custom/" prefix, for example (which could be desirable for certain users).  I think it probably makes sense just to drop down to just having the prefix specified.